### ISSUE TITLE
Error on image extraction (don't know how to __str__ this object)

### ISSUE THREAD

**Issue by akhmerov** [pikepdf/pikepdf#108]
I've encountered a failure trying to extract images from a PDF. Unfortunately the original file has private data, and was produced externally. I am, however, able to narrow down the failure to a specific part of the PDF:

```
> .../lib/python3.8/site-packages/pikepdf/models/image.py(53)<listcomp>()
     51         items = [item for item in value]
     52         if len(items) == 4 and items[0] == '/Indexed':
---> 53             result = [str(items[n]) for n in range(3)]
     54             result.append(bytes(items[3]))
     55             return result

ipdb> items                                                                                                                                                                                                
[pikepdf.Name("/Indexed"), pikepdf.Array([ "/ICCBased", None ]), 64, pikepdf.String("''L-++T@?{IJo))>?@`UV†km£†…ÄŸŽæE;a'6/LPEqk\\Œ–s»£¢±**.BAHYXappz‹⁄ﬂ�ž­·µÇÎÍàåäùS`^)0/?IGTa_izw~ﬁ’ﬂ«¨©ÄÀ¾ÜØÓõð7&8N5OQCP 09SERmZk⁄o–€—žºı·Ó¯ÐíÄêww]..$JJ:ggO……e��{»»‚××§óó½ÿÿÿ")]
```

Please let me know if any additional input would be useful.

<details>
<summary>full traceback</summary>

```
~/Bot/coding/zesje/zesje/image_extraction.py in extract_image_pikepdf(page)
    272         if abs(pdf_ratio - image_ratio) > 0.03 * pdf_ratio:
    273             raise ValueError('Image has incorrect dimensions')
--> 274         return pdf_image.as_pil_image()
    275 
    276 

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in as_pil_image(self)
    570             pass
    571 
--> 572         im = self._extract_transcoded()
    573         if not im:
    574             raise UnsupportedImageTypeError(repr(self))

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in _extract_transcoded(self)
    412 
    413         im = None
--> 414         if self.mode == 'RGB' and self.bits_per_component == 8:
    415             # No point in accessing the buffer here, size qpdf decodes to 3-byte
    416             # RGB and Pillow needs RGBX for raw access

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in mode(self)
    180 
    181         m = ''
--> 182         if self.indexed:
    183             m = 'P'
    184         elif self.bits_per_component == 1:

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in indexed(self)
    164     def indexed(self):
    165         """``True`` if the image has a defined color palette"""
--> 166         return '/Indexed' in self._colorspaces
    167 
    168     @property

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in _colorspaces(self)
    113     def _colorspaces(self):
    114         """Colorspace (low-level)"""
--> 115         return self._metadata('ColorSpace', array_str_colorspace, [])
    116 
    117     @property

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in _metadata(self, name, type_, default)
    335 
    336     def _metadata(self, name, type_, default):
--> 337         return metadata_from_obj(self.obj, name, type_, default)
    338 
    339     @property

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in metadata_from_obj(obj, name, type_, default)
     75     val = getattr(obj, name, default)
     76     try:
---> 77         return type_(val)
     78     except TypeError:
     79         if val is None:

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in array_str_colorspace(value)
     51         items = [item for item in value]
     52         if len(items) == 4 and items[0] == '/Indexed':
---> 53             result = [str(items[n]) for n in range(3)]
     54             result.append(bytes(items[3]))
     55             return result

~/miniconda3/envs/zesje-dev/lib/python3.8/site-packages/pikepdf/models/image.py in <listcomp>(.0)
     51         items = [item for item in value]
     52         if len(items) == 4 and items[0] == '/Indexed':
---> 53             result = [str(items[n]) for n in range(3)]
     54             result.append(bytes(items[3]))
     55             return result

NotImplementedError: don't know how to __str__ this object
```
</details>

**Comment by jbarlow83** [comment]
That's a bug - fixed for v1.17.0.

Didn't check for the combined case of an image that uses a color profile and a color palette.

**Comment by akhmerov** [comment]
I confirm, this is fixed in v1.17.0!

Thanks a lot!
